### ISSUE TITLE
SAML2DefaultResponseValidator wantAssertionsSigned cannot evaluate to true

### ISSUE THREAD

**Issue by jhegg** [pac4j/pac4j#502]
Hello,

On the 1.8.x branch, when a forged SAML response received by the SP (using pac4j-saml) has the IdP signature removed, the `SAML2DefaultResponseValidator#validateAssertionSignature` method is supposed to fall back to checking whether `wantAssertionsSigned == true`. In our testing, the `context.getSPSSODescriptor()` call always returns null, and per the next line the boolean is always set to `false` (see: https://github.com/pac4j/pac4j/blob/417ea36c108efa878acfceebc535dd7ffee372b5/pac4j-saml/src/main/java/org/pac4j/saml/sso/impl/SAML2DefaultResponseValidator.java#L653-L657). This means that you can forge any assertion, such as the NameId, and it is treated as valid.

For example, this SAML response (with some minor redaction of names/URLs) had the username modified to `administrator` to simulate a privilege escalation attack:

```xml
<samlp:Response Version="2.0" ID="nHk6INJVcukHphdvDwYnyNmFt.f" IssueInstant="2016-02-24T00:52:08.029Z" InResponseTo="_rmwurbdkpt0o0ao4flcvujpxu569eusoovpj34r"
    xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol">
    <saml:Issuer
        xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion">saml-dev
    </saml:Issuer>
    <samlp:Status>
        <samlp:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Success"/>
    </samlp:Status>
    <saml:Assertion ID="aj_HKuHiEkKZt2era2IXQ46f2Ib" IssueInstant="2016-02-24T00:52:08.037Z" Version="2.0"
        xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion">
        <saml:Issuer>saml-dev</saml:Issuer>
        <saml:Subject>
            <saml:NameID Format="urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified">administrator</saml:NameID>
            <saml:SubjectConfirmation Method="urn:oasis:names:tc:SAML:2.0:cm:bearer">
                <saml:SubjectConfirmationData Recipient="https://www.example.com" NotOnOrAfter="2016-02-24T00:57:08.037Z" InResponseTo="_rmwurbdkpt0o0ao4flcvujpxu569eusoovpj34r"/>
            </saml:SubjectConfirmation>
        </saml:Subject>
        <saml:Conditions NotBefore="2016-02-24T00:47:08.037Z" NotOnOrAfter="2016-02-24T00:57:08.037Z">
            <saml:AudienceRestriction>
                <saml:Audience>my-entity-id</saml:Audience>
            </saml:AudienceRestriction>
        </saml:Conditions>
        <saml:AuthnStatement SessionIndex="aj_HKuHiEkKZt2era2IXQ46f2Ib" AuthnInstant="2016-02-24T00:52:08.034Z">
            <saml:AuthnContext>
                <saml:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:unspecified</saml:AuthnContextClassRef>
            </saml:AuthnContext>
        </saml:AuthnStatement>
        <saml:AttributeStatement>
            <saml:Attribute Name="SAML_MEMBEROF" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
                <saml:AttributeValue xsi:type="xs:string"
                    xmlns:xs="http://www.w3.org/2001/XMLSchema"
                    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">CN=Server Operators,CN=Builtin,DC=saml-dev,DC=example,DC=com
                </saml:AttributeValue>
                <saml:AttributeValue xsi:type="xs:string"
                    xmlns:xs="http://www.w3.org/2001/XMLSchema"
                    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">CN=Administrators,CN=Builtin,DC=saml-dev,DC=example,DC=com
                </saml:AttributeValue>
            </saml:Attribute>
            <saml:Attribute Name="SAML_EMAIL" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
                <saml:AttributeValue xsi:type="xs:string"
                    xmlns:xs="http://www.w3.org/2001/XMLSchema"
                    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">email@example.com
                </saml:AttributeValue>
            </saml:Attribute>
        </saml:AttributeStatement>
    </saml:Assertion>
</samlp:Response>

```

Is there a bug here, or are we configuring something incorrectly? We followed the pattern of the j2e-pac4j project, and we have file-based SP metadata. It doesn't appear that we can make any kind of configuration change in our code to affect that `wantAssertionsSigned` variable to always be `true` (when the SAML response does not have a signature).

I was also confused by what the validator code is doing, in that it seems to be asking the `SAML2MessageContext` for its `SPSSODescriptor`, and then getting the `wantAssertionsSigned` variable from that. I'm not sure how the `Saml2Client` and its metadata configuration gets wired up to the `SAML2MessageContext`, since that seems to be built based on the inbound HTTP request and SAML payload. I'd rather have some configuration item in `SAML2ClientConfiguration` that I could toggle, although realistically for my usages I always want to enforce signing on the inbound SAML payload (whether from a real IdP or a forged one).

**Comment by leleuj** [comment]
@mmoayyed will know better...

**Comment by mmoayyed** [comment]
So two things first:

1. You need to investigate why the SSODescriptor is null. Trace the context down. 
2. I think we could allow a default global option for "wantAssertionsSigned" such that if the SSODescriptor is null, the global option takes over. 

But I would want to know why the descriptor is null to begin with.

**Comment by jhegg** [comment]
Thanks @mmoayyed . What we're seeing is that in `SAML2WebSSOMessageReceiver#receiveMessage` the `context` parameter passed to the method has the valid SSODescriptor, but the `decodedCtx` variable that is passed to the validator has a null SSODescriptor. I don't think I'm fully understanding what's going on, but it seems the `decodedCtx` is effectively copying pieces of the original `context` parameter, so presumably it is not copying the SSODescriptor into the `decodedCtx`. I'm attempting to write a test to verify.

**Comment by mmoayyed** [comment]
Thanks. Keep us posted please.

**Comment by jhegg** [comment]
I'm having problems writing a test for `SAML2WebSSOMessageReceiver#receiveMessage` because it requires a lot of mocking to simulate the original context. In parallel, my coworker has tested with a small change to copy the SPSSODescriptor to the `decodedCtx`, like this:

```java
decodedCtx.getSAMLSelfMetadataContext().setRoleDescriptor(context.getSPSSODescriptor());
```

I don't know enough to understand if this is the correct fix, but it appears to resolve the issue of the SPSSODescriptor being null from within the `SAML2DefaultResponseValidator#validateAssertionSignature`.

Does that seem like a reasonable fix? Should we also change the default fallback of `wantAssertionsSigned` from `false` to instead be `true`, so that it securely falls back when SPSSODescriptor is null? I understand that may be a behavior change for other users, although I don't know why anyone would want unsigned assertions (except for during testing?).

**Comment by mmoayyed** [comment]
>I don't know enough to understand if this is the correct fix, but it appears to resolve the issue of the SPSSODescriptor being null from within the SAML2DefaultResponseValidator#validateAssertionSignature.

Seems reasonable from the looks of it, but I'll do a quick pass to make sure. Stay tuned. 

> Should we also change the default fallback of wantAssertionsSigned from false to instead be true, so that it securely falls back when SPSSODescriptor is null?

Yes, but I prefer if we make that optional. We could expose an option in the saml config that forces the default behavior of that flag, and we could by default set it to true. If anyone wanted it off, they could just simply tweak that config by a simple method call on the config.

**Comment by mmoayyed** [comment]
To elaborate:

1. I think the initial pass at SAML features of pac4j were tested against an SP that did not want signed assertions for whatever reason. So the behavior was kept as false by default. Possibly a bad idea. 
2. Setting it to true, MIGHT break some integrations with other SPs but hopefully it will be a realization point where the dev starts to understand that it's likely a bad idea to turn that but off, and it would lead to further conversations :) I like providing choices when it's at no cost to the platform. The option might seem like a bad idea, but I prefer to let developers make that choice.

**Comment by jhegg** [comment]
@mmoayyed Adding the option sounds great and I'll start working up a pull request that uses the option as a fallback if SPSSODescriptor is ever null, thanks!

I can also work up a pull request for the proposed fix of `decodedCtx.getSAMLSelfMetadataContext().setRoleDescriptor(context.getSPSSODescriptor())`, unless you are already working on that as part of researching my findings?

**Comment by mmoayyed** [comment]
I was just trying to verify, and it looks like it's a reasonable fix. You're due all the credit :) Go ahead with the PR. Thanks much! Look fwd to it.

**Comment by jhegg** [comment]
@mmoayyed Great, thanks! Working on it now.

**Comment by mmoayyed** [comment]
One additional note, that you should likely:

1. Post 2 PRs, one for 1.8.x and one for master.
2. One for master, and @leleuj or myself could cherry pick. 

Whatever works.

**Comment by jhegg** [comment]
Thanks, I can post PRs for both.

Is there any interest in getting the 1.8.x branch merged into master (null/empty merge or whatever), so that new changes can be more easily merged into master, instead of cherry picking?

**Comment by mmoayyed** [comment]
I'd leave that for Jérôme, but likely, it's not so easily done. For one thing, there are some heavy changes in master now that a merge/rebase would a lot more work than cherry picking.

**Comment by leleuj** [comment]
Yes, the master branch has strongly changed from the 1.8.x branch so a merge won't work easily.

**Comment by leleuj** [comment]
Excellent contribution! Many thanks.

I plan a 1.8.7 release at the beginning of March.
